core/corehttp/gateway_handler.go

```diff
@@ -199,8 +199,15 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 		webError(w, "ipfs resolve -r "+escapedURLPath, err, http.StatusServiceUnavailable)
 		return
 	default:
-		webError(w, "ipfs resolve -r "+escapedURLPath, err, http.StatusNotFound)
-		return
+		resolved404Path, err404 := i.searchUpTreeFor404(r, parsedPath)
+		if err404 != nil {
+			webError(w, "ipfs resolve -r "+escapedURLPath, err, http.StatusNotFound)
+			return
+		}
+
+		w.WriteHeader(http.StatusNotFound)
```

Comment:
* This will cause us to try to write the status multiple times. Go doesn't like that.
* This will drop any further headers and log a bunch of errors.

We should probably set an "is404" boolean, and check it later.

Reply:
Also, if 404.html happens to be a directory, strange things will happen.